src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -7,400 +7,441 @@
-        case dp: DefaultPromise[_] =>
-          compressedRoot(dp).link(target)
-        case listeners: List[_] if compareAndSet(listeners, target) =>
-          if (listeners.nonEmpty)
-            listeners.asInstanceOf[List[CallbackRunnable[T]]].foreach(target.dispatchOrAddCallback(_))
-        case _ =>
-          link(target)
+        } else if (state.isInstanceOf[Link[T]]) {
+          // TODO: does relinking ever make sense?
+          state.asInstanceOf[Link[T]].relink(link = state.asInstanceOf[Link[T]], target = target, owner = this)
+        } else /*if (state.isInstanceOf[Callbacks[T]]) */ {
+          val l = if (link ne null) link else new Link(target)
+          val p = l.promise(this)
+          if (p ne this) {
+            if (compareAndSet(state, l)) {
+              if (state ne Noop) p.dispatchOrAddCallbacks(p.get(), state.asInstanceOf[Callbacks[T]])
+            } else linkRootOf(p, l)
+          }
+        }
       }
-    }
   }
 
-  /** An already completed Future is given its result at creation.
-   *
-   *  Useful in Future-composition when a value to contribute is already available.
+  // Constant byte tags for unpacking transformation function inputs or outputs
+  final val Xform_noop          = 0
+  final val Xform_map           = 1
+  final val Xform_flatMap       = 2
+  final val Xform_transform     = 3
+  final val Xform_transformWith = 4
+  final val Xform_foreach       = 5
+  final val Xform_onComplete    = 6
+  final val Xform_recover       = 7
+  final val Xform_recoverWith   = 8
+  final val Xform_filter        = 9
+  final val Xform_collect       = 10
+  //final val Xform_failed        = 11
+
+    /* Marker trait
    */
-  object KeptPromise {
-    import scala.concurrent.Future
-    import scala.reflect.ClassTag
+  sealed trait Callbacks[-T] {
+    def submitWithValue(v: Try[T]): Unit
+  }
+
+  final class ManyCallbacks[-T](final val first: Callbacks[T], final val last: Callbacks[T]) extends Callbacks[T] {
+    // NOTE: This does grow the stack for *linked* callbacks which are transported across the links,
+    // these should normally be rather flat.
+    final def submitWithValue(v: Try[T]): Unit = {
+      first.submitWithValue(v)
+      last.submitWithValue(v)
+    }
+    override final def toString: String = "ManyCallbacks"
+  }
 
-    private[this] sealed trait Kept[T] extends Promise[T] {
-      def result: Try[T]
+  private[this] final val Noop = new Transformation[Nothing, Nothing](Xform_noop, null, InternalCallbackExecutor)
+
+  final class Transformation[-F, T] private[this] (
+    private[this] final var _fun: Any => Any,
+    private[this] final var _arg: AnyRef,
+    private[this] final val _xform: Byte
+  ) extends DefaultPromise[T]() with Callbacks[F] with Runnable with OnCompleteRunnable {
+    final def this(xform: Int, f: _ => _, ec: ExecutionContext) = this(f.asInstanceOf[Any => Any], ec.prepare(): AnyRef, xform.asInstanceOf[Byte])
```

Comment:
It would be nice to have a comment explaining why using an untyped function is required, (as well as what a `Transformation` is in general). I don't like giving up entirely on type safety without good reason - and I'm not doubting that there *is* a good reason - but it would be helpful to know what that is.

Reply:
@NthPortal Will add documentation. I had several drafts which preserved as much type information as possible, but it quickly became either too expensive (memory or indirection) or too unwieldly (parametrization). I can try with a type member approach to see if that works out better.

In general Transformation is a structure which encodes its own dispatch table to control its monomorphicity (inlineability, cheap type tests etc)